ykman/piv.py

```diff
@@ -814,14 +815,27 @@ def sign_raw(self, slot, algorithm, message):
         return self._raw_sign_decrypt(slot, algorithm, Tlv(0x81, message),
                                       _sign_len_conditions[algorithm])
 
-    def sign(self, slot, algorithm, message):
+    def sign(self, slot, algorithm, message, pin=None, touch_callback=None):
+        message_hash = None
+
         if algorithm in (ALGO.RSA1024, ALGO.RSA2048):
-            message = _pkcs1_15_pad(algorithm, message)
+            message_hash = _pkcs1_15_pad(algorithm, message)
         elif algorithm in (ALGO.ECCP256, ALGO.ECCP384):
             h = hashes.Hash(hashes.SHA256(), default_backend())
             h.update(message)
-            message = h.finalize()
-        return self.sign_raw(slot, algorithm, message)
+            message_hash = h.finalize()
+        try:
+            return self.sign_raw(slot, algorithm, message_hash)
+        except APDUError as e:
+            if e.sw == SW.ACCESS_DENIED:
+                if pin:
+                    self.verify(pin, touch_callback=touch_callback)
+                    return self.sign(slot, algorithm, message, pin=pin)
+                else:
```

Comment:
You can get rid of this else-clause and just de-indent the raise at the bottom of the method by one level.

Reply:
Yes. Is that the require style in this codebase? Personally I generally prefer explicit branches over early returns.

Reply:
I prefer to avoid else branches [if possible].(https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement?answertab=votes#tab-top)